sxt/proof/sumcheck/polynomial_utility.cc

```diff
@@ -0,0 +1,93 @@
+/** Proofs GPU - Space and Time's cryptographic proof algorithms on the CPU and GPU.
+ *
+ * Copyright 2024-present Space and Time Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "sxt/proof/sumcheck/polynomial_utility.h"
+
+#include <cassert>
+
+#include "sxt/scalar25/operation/add.h"
+#include "sxt/scalar25/operation/mul.h"
+#include "sxt/scalar25/operation/muladd.h"
+#include "sxt/scalar25/operation/sub.h"
+#include "sxt/scalar25/type/element.h"
+
+namespace sxt::prfsk {
+//--------------------------------------------------------------------------------------------------
+// sum_polynomial_01
+//--------------------------------------------------------------------------------------------------
+void sum_polynomial_01(s25t::element& e, basct::cspan<s25t::element> polynomial) noexcept {
+  if (polynomial.empty()) {
+    e = s25t::element{};
+    return;
+  }
+  e = polynomial[0];
+  for (unsigned i = 1; i < polynomial.size(); ++i) {
+    s25o::add(e, e, polynomial[i]);
+  }
+}
+
+//--------------------------------------------------------------------------------------------------
+// evaluate_polynomial
+//--------------------------------------------------------------------------------------------------
+void evaluate_polynomial(s25t::element& e, basct::cspan<s25t::element> polynomial,
+                         const s25t::element& x) noexcept {
+  if (polynomial.empty()) {
+    e = s25t::element{};
+    return;
+  }
+  auto i = polynomial.size();
+  --i;
+  e = polynomial[i];
+  while (i > 0) {
+    --i;
+    s25o::muladd(e, e, x, polynomial[i]);
+  }
+}
+
+//--------------------------------------------------------------------------------------------------
+// expand_products
+//--------------------------------------------------------------------------------------------------
+CUDA_CALLABLE
+void expand_products(basct::span<s25t::element> p, const s25t::element* mles, unsigned n,
+                     unsigned step, basct::cspan<unsigned> terms) noexcept {
+  auto num_terms = terms.size();
+  assert(num_terms > 0 && p.size() == num_terms + 1u);
```

Comment:
Is `n` just an index multiplier? Should there be a check in the `assert` to ensure a specific relationship between `n`, `step` and `num_terms`?

Reply:
I added an assert. The requirement would be that n > step.